### PR TITLE
MOSIP-11739 : added workflowInstanceId changes

### PR DESCRIPTION
**Pull request by MonobikashDas**

### CODE REVIEW

File: registration-processor/pre-processor/registration-processor-securezone-notification-stage/src/main/java/io/mosip/registration/processor/securezone/notification/stage/SecurezoneNotificationStage.java [hunk 1]
```diff
@@ -279,4 +307,37 @@ public void sendMessage(MessageDTO messageDTO) {
...
+        // find all records for same additionalInfoReqId.
+        List<SyncRegistrationEntity> entities = syncRegistrationService.findByAdditionalInfoReqId(entity.getAdditionalInfoReqId());
+        // if multiple records are present for same additionalInfoReqId then check in registration table how many packets are received
+        if (!CollectionUtils.isEmpty(entities) && entities.size() > 1) {
+            List<String> workflowInstanceIds = entities.stream().map(e -> e.getWorkflowInstanceId()).collect(Collectors.toList());
+            List<InternalRegistrationStatusDto> dtos = new ArrayList<>();
+            for (String w : workflowInstanceIds) {
+                InternalRegistrationStatusDto dto = registrationStatusService.getRegistrationStatus(
+                        messageDTO.getRid(), messageDTO.getReg_type(), messageDTO.getIteration(), w);
+                if (dto != null )
```
Comment (vishwa-vyom) [inline registration-processor/pre-processor/registration-processor-securezone-notification-stage/src/main/java/io/mosip/registration/processor/securezone/notification/stage/SecurezoneNotificationStage.java:323]: message DTO anyways contains rid, reg type and iteration, can we just directly query registration table for all matches sorted by create date time asc, if the current workflow Id is the first one then we will say isValid as true else false ?

File: registration-processor/pre-processor/registration-processor-securezone-notification-stage/src/test/java/io/mosip/registration/processor/securezone/notification/stage/SecurezoneNotificationStageTest.java [hunk 2]
```diff
@@ -430,7 +430,7 @@ public void processTest() {
 
     @Test
     public void dbExceptionTest() {
-        Mockito.when(registrationStatusService.getRegistrationStatus(anyString(), any(), any())).thenThrow(new TablenotAccessibleException("exception"));
+        Mockito.when(registrationStatusService.getRegistrationStatus(anyString(), any(), any(), any())).thenThrow(new TablenotAccessibleException("exception"));
```
Comment (vishwa-vyom) [inline registration-processor/pre-processor/registration-processor-securezone-notification-stage/src/test/java/io/mosip/registration/processor/securezone/notification/stage/SecurezoneNotificationStageTest.java:433]: Lets also add a test case for the duplicate validation

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]

**Review by vishwa-vyom — APPROVED** [review]
Since there is dependency on these changes by other developers, I am conditional approving this PR and expecting all the review comments of this PR to be taken care as a separate PR.

File: registration-processor/core-processor/registration-processor-manual-verification-stage/src/main/java/io/mosip/registration/processor/manual/verification/service/ManualVerificationService.java [hunk 3]
```diff
@@ -1,74 +1,50 @@
...
-	 *            The registration ID
-	 * @param fileName
-	 *            The file required
-	 * @return The file as bytes
-	 * @throws java.io.IOException 
-	 * @throws IOException 
-	 * @throws ApisResourceAccessException 
-	 */
-	public byte[] getApplicantFile(String regId, String fileName, String source) throws IOException, ApisResourceAccessException, PacketManagerException, JsonProcessingException;
-
```
[the remark this answers is not in the document]
Comment (MonobikashDas) [reply registration-processor/core-processor/registration-processor-manual-verification-stage/src/main/java/io/mosip/registration/processor/manual/verification/service/ManualVerificationService.java:50]: its not being used.

File: registration-processor/init/registration-processor-packet-receiver-stage/src/main/java/io/mosip/registration/processor/packet/receiver/service/impl/PacketReceiverServiceImpl.java [hunk 4]
```diff
@@ -142,6 +141,7 @@ public MessageDTO validatePacket(File file, String stageName) {
 			regEntity = syncRegistrationService.findByPacketId(packetId);
 			String registrationId = regEntity.getRegistrationId();
 			messageDTO.setRid(registrationId);
+			messageDTO.setWorkflowInstanceId(regEntity.getPacketId());
```
[the remark this answers is not in the document]
Comment (MonobikashDas) [reply registration-processor/init/registration-processor-packet-receiver-stage/src/main/java/io/mosip/registration/processor/packet/receiver/service/impl/PacketReceiverServiceImpl.java:144]: this could be because of code rebase. I corrected it and pushed code directly since its required.

File: registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/service/impl/SubWorkflowMappingServiceImpl.java [hunk 5]
```diff
@@ -1,69 +1,70 @@
...
+    @Autowired
+    private BaseRegProcRepository<SubWorkflowMappingEntity, String> subWorkflowRepository;
+
+    @Override
+    public List<SubWorkflowDto> getWorkflowMappingByReqId(String infoRequestId) {
+        List<SubWorkflowMappingEntity> subWorkflowMappingEntity = subWorkflowRepository.workflowMappingByReqId(infoRequestId);
+        List<SubWorkflowDto> subWorkflowDtos = new ArrayList<>();
+        if (CollectionUtils.isNotEmpty(subWorkflowMappingEntity))
+            subWorkflowMappingEntity.forEach(m -> subWorkflowDtos.add(new SubWorkflowDto(m.getId().getRegId(),
+                    m.getId().getAdditionalInfoReqId(), m.getProcess(), m.getIteration(),
```
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/service/impl/SubWorkflowMappingServiceImpl.java:26]: This part I have taken care as part of https://github.com/mosip/registration/pull/1054 PR

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/WorkflowInternalActionVerticle.java [hunk 6]
```diff
@@ -1,434 +1,435 @@
...
+			updateAudit(description, isTransactionSuccessful, registrationId);
+		}
+		return object;
+
+	}
+
+	private void processMarkAsReprocess(WorkflowInternalActionDTO workflowInternalActionDTO) {
+		InternalRegistrationStatusDto registrationStatusDto = registrationStatusService
+				.getRegistrationStatus(workflowInternalActionDTO.getRid(), workflowInternalActionDTO.getReg_type(),
+						workflowInternalActionDTO.getIteration(), null);
```
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/WorkflowInternalActionVerticle.java:218]: This is also taken care as part of https://github.com/mosip/registration/pull/1052 PR

File: registration-processor/pre-processor/registration-processor-packet-validator-stage/src/main/java/io/mosip/registration/processor/stages/packet/validator/PacketValidateProcessor.java [hunk 7]
```diff
@@ -449,8 +449,8 @@ public MessageDTO process(MessageDTO object, String stageName) {
 
 	}
 
-	private boolean isValidSupervisorStatus() {
-		SyncRegistrationEntity regEntity = syncRegistrationService.findByRegistrationId(registrationId);
```
[the remark this answers is not in the document]
Comment (MonobikashDas) [reply registration-processor/pre-processor/registration-processor-packet-validator-stage/src/main/java/io/mosip/registration/processor/stages/packet/validator/PacketValidateProcessor.java:453]: its already fixed. I can see latest code -
private boolean isValidSupervisorStatus(MessageDTO messageDTO) {
		SyncRegistrationEntity regEntity = syncRegistrationService.findByWorkflowInstanceId(messageDTO.getWorkflowInstanceId());
		if (regEntity.getSupervisorStatus().equalsIgnoreCase(APPROVED)) {
			return true;

		} else if (regEntity.getSupervisorStatus().equalsIgnoreCase(REJECTED)) {
			return false;
		}
		return false;
	}

File: registration-processor/init/registration-processor-packet-receiver-stage/src/main/java/io/mosip/registration/processor/packet/receiver/service/impl/PacketReceiverServiceImpl.java [hunk 8]
```diff
@@ -545,13 +548,11 @@ public MessageDTO processPacket(File file) {
...
 	}
 
-	private int getIterationFromPacketName(String packetId, SyncRegistrationEntity regEntity) {
-        if (mainProcess.contains(regEntity.getRegistrationType()))
-            return 1;
-        else {
-            char c = packetId.charAt(packetId.length() - 1);
-            return Integer.parseInt(String.valueOf(c));
-        }
+	private int getIterationFromPacketName(SyncRegistrationEntity regEntity) {
```
[the remark this answers is not in the document]
Comment (MonobikashDas) [reply registration-processor/init/registration-processor-packet-receiver-stage/src/main/java/io/mosip/registration/processor/packet/receiver/service/impl/PacketReceiverServiceImpl.java:552]: changed to getIterationForSyncRecord()

File: registration-processor/init/registration-processor-packet-receiver-stage/src/main/java/io/mosip/registration/processor/packet/receiver/service/impl/PacketReceiverServiceImpl.java [hunk 9]
```diff
@@ -243,10 +243,11 @@ private void validatePacketWithSync(SyncRegistrationEntity regEntity, String reg
...
 	 *             Signals that an I/O exception has occurred.
 	 */
-	private boolean storePacket(String packetId, String stageName, SyncRegistrationEntity regEntity, InternalRegistrationStatusDto dto,
-			LogDescription description) throws IOException {
+	private boolean storePacket(String stageName, SyncRegistrationEntity regEntity, InternalRegistrationStatusDto dto,
+			LogDescription description) {
 		Boolean storageFlag = false;
-		dto = registrationStatusService.getRegistrationStatus(regEntity.getRegistrationId(), regEntity.getRegistrationType(), null);
+		dto = registrationStatusService.getRegistrationStatus(regEntity.getRegistrationId(),
+				regEntity.getRegistrationType(), 1, regEntity.getWorkflowInstanceId());
```
[the remark this answers is not in the document]
Comment (MonobikashDas) [reply registration-processor/init/registration-processor-packet-receiver-stage/src/main/java/io/mosip/registration/processor/packet/receiver/service/impl/PacketReceiverServiceImpl.java:250]: calling method getIterationForSyncRecord()

File: registration-processor/pre-processor/registration-processor-securezone-notification-stage/src/main/java/io/mosip/registration/processor/securezone/notification/stage/SecurezoneNotificationStage.java (hunk 1 above)
Comment (MonobikashDas) [reply registration-processor/pre-processor/registration-processor-securezone-notification-stage/src/main/java/io/mosip/registration/processor/securezone/notification/stage/SecurezoneNotificationStage.java:323]: first i am getting all the workflowInstanceId for same additional info req id.
then i get the records in registration table for those workflowInstanceId
then i check datetime on records of registration table
so if packet is uploaded only then record will be present in registration table

File: registration-processor/registration-processor-info-storage-service/src/main/java/io/mosip/registration/processor/packet/storage/entity/ManualVerificationPKEntity.java [hunk 10]
```diff
@@ -1,94 +1,105 @@
...
+
+	/** The matched ref id. */
+	@Column(name = "matched_ref_id")
+	private String matchedRefId;
+
+	/** The matched ref type. */
+	@Column(name = "matched_ref_type")
+	private String matchedRefType;
+
+	@Column(name = "workflow_instance_id")
```
[the remark this answers is not in the document]
Comment (MonobikashDas) [reply registration-processor/registration-processor-info-storage-service/src/main/java/io/mosip/registration/processor/packet/storage/entity/ManualVerificationPKEntity.java:32]: done

File: registration-processor/registration-processor-info-storage-service/src/main/java/io/mosip/registration/processor/packet/storage/entity/RegLostUinDetPKEntity.java [hunk 11]
```diff
@@ -14,6 +14,9 @@
 	@Column(name = "reg_id")
```
[the remark this answers is not in the document]
Comment (MonobikashDas) [reply registration-processor/registration-processor-info-storage-service/src/main/java/io/mosip/registration/processor/packet/storage/entity/RegLostUinDetPKEntity.java:14]: internally its calling RegistrationStatusDao.find method which we have already discussed is a big change. So lets do it later.

File: registration-processor/registration-processor-info-storage-service/src/main/java/io/mosip/registration/processor/packet/storage/utils/ABISHandlerUtil.java [hunk 12]
```diff
@@ -72,11 +72,11 @@
...
 	public Set<String> getUniqueRegIds(String registrationId, String registrationType, 
-										int iteration, ProviderStageName stageName) throws ApisResourceAccessException, JsonProcessingException, PacketManagerException, IOException {
+										int iteration, String workflowInstanceId, ProviderStageName stageName) throws ApisResourceAccessException, JsonProcessingException, PacketManagerException, IOException {
 		regProcLogger.debug(LoggerFileConstant.SESSIONID.toString(), LoggerFileConstant.USERID.toString(),
 				registrationId, "ABISHandlerUtil::getUniqueRegIds()::entry");
 		
-		String latestTransactionId = utilities.getLatestTransactionId(registrationId, registrationType, iteration);
+		String latestTransactionId = utilities.getLatestTransactionId(registrationId, registrationType, iteration, workflowInstanceId);
 
 		List<String> regBioRefIds = packetInfoDao.getAbisRefMatchedRefIdByRid(registrationId);
```
[the remark this answers is not in the document]
Comment (MonobikashDas) [reply registration-processor/registration-processor-info-storage-service/src/main/java/io/mosip/registration/processor/packet/storage/utils/ABISHandlerUtil.java:81]: internally its calling RegistrationStatusDao.find method which we have already discussed is a big change. So lets do it later.

File: registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/service/impl/SyncRegistrationServiceImpl.java [hunk 13]
```diff
@@ -490,12 +489,21 @@ private boolean validateRegistrationID(SyncRegistrationDto registrationDto,
...
 			syncRegistration.setCreateDateTime(LocalDateTime.now(ZoneId.of("UTC")));
-			syncRegistration.setId(RegistrationUtility.generateId());
+			syncRegistration.setWorkflowInstanceId(RegistrationUtility.generateId());
 			syncRegistrationDao.save(syncRegistration);
 			syncResponseDto.setRegistrationId(registrationDto.getRegistrationId());
 
 			eventId = EventId.RPR_407.toString();
 		}
+
+		syncRegistration = convertDtoToEntity(registrationDto, referenceId, timeStamp);
```
[the remark this answers is not in the document]
Comment (MonobikashDas) [reply registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/service/impl/SyncRegistrationServiceImpl.java:499]: i didnot understand. lets discuss later.

File: registration-processor/init/registration-processor-packet-receiver-stage/src/main/java/io/mosip/registration/processor/packet/receiver/service/impl/PacketReceiverServiceImpl.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply registration-processor/init/registration-processor-packet-receiver-stage/src/main/java/io/mosip/registration/processor/packet/receiver/service/impl/PacketReceiverServiceImpl.java:250]: I see the same method getIterationForSyncRecord(regEntity) is called twice in the same method, can we store the value in a variable and reuse ? This is to avoid 2 dips to DB

File: registration-processor/registration-processor-info-storage-service/src/main/java/io/mosip/registration/processor/packet/storage/entity/RegBioRefEntity.java [hunk 14]
```diff
@@ -33,6 +33,9 @@
 	@Column(name = "upd_dtimes")
 	private LocalDateTime updDtimes;
 
+	@Column(name = "workflow_instance_id")
```
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply registration-processor/registration-processor-info-storage-service/src/main/java/io/mosip/registration/processor/packet/storage/entity/RegBioRefEntity.java:36]: I see the crDtimes is moved out of PK entity, this means that will not do a new insert to this table when the reprocessing is done ?

File: registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/entity/BaseRegistrationPKEntity.java [hunk 15]
```diff
@@ -24,6 +24,9 @@
 
 	@Column
 	private Integer iteration=1;
+
+	@Column(name = "workflow_instance_id")
+	protected String workflowInstanceId;
```
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/entity/BaseRegistrationPKEntity.java:29]: I see this is also taken care, correct ?

File: registration-processor/registration-processor-info-storage-service/src/main/java/io/mosip/registration/processor/packet/storage/utils/ABISHandlerUtil.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply registration-processor/registration-processor-info-storage-service/src/main/java/io/mosip/registration/processor/packet/storage/utils/ABISHandlerUtil.java:81]: Lets do a quick discusson on the change
